zephyr/lib/pacovr.c

```diff
@@ -0,0 +1,253 @@
+	p->batch_size = batch_size;
+	p->pages = pages;
+	p->batch_ptr = p->base + scratch_size;
+	p->batch_free_count = 0;
+
+	/* init scratch heap */
+	sys_heap_init(&p->scratch.heap, p->base, scratch_size);
+
+	LOG_INF("pacovr created at base %p total size 0x%x pages %d scratch 0x%x batch 0x%x",
+		p->base, total_size, pages, scratch_size, batch_size);
+
+	return p;
+}
+
+/**
+ * @brief Destroy a PACOVR instance.
+ *
+ * @param p Pointer to the PACOVR instance to destroy.
+ */
+void pacovr_destroy(struct pacovr *p)
+{
+	if (!p)
+		return;
+	LOG_INF("pacovr destroy base %p size 0x%x pages %d batch used 0x%x free count %d",
+		p->base, p->size, p->pages, p->batch_used, p->batch_free_count);
+	vpage_free(p->base);
+	rfree(p);
+}
+
+/**
+ * @brief Allocate memory from the PACOVR scratch heap.
+ *
+ * @param p Pointer to the PACOVR instance.
+ * @param size Size of the allocation.
+ * @return void* Pointer to the allocated memory, or NULL on failure.
+ */
+void *pacovr_scratch_alloc(struct pacovr *p, size_t size)
+{
+	void *ptr;
+
+	if (!p || !size)
+		return NULL;
+
+	ptr = sys_heap_alloc(&p->scratch.heap, size);
+	if (!ptr) {
+		LOG_ERR("error: pacovr scratch alloc failed");
+		return NULL;
+	}
+
+	return ptr;
+}
+
+/**
+ * @brief Allocate memory with alignment from the PACOVR scratch heap.
+ *
+ * @param p Pointer to the PACOVR instance.
+ * @param size Size of the allocation.
+ * @param align Alignment of the allocation.
+ * @return void* Pointer to the allocated memory, or NULL on failure.
+ */
+void *pacovr_scratch_alloc_align(struct pacovr *p, size_t size, size_t align)
+{
+	void *ptr;
+
+	if (!p || !size)
+		return NULL;
+
+	/* align up size to 4 bytes - force aligned loads and stores */
+	if (!align)
+		align = sizeof(uint32_t);
+
+	ptr = sys_heap_aligned_alloc(&p->scratch.heap, size, align);
+	if (!ptr) {
+		LOG_ERR("error: pacovr scratch alloc failed");
+		return NULL;
+	}
+
+	return ptr;
+}
+
+/**
+ * @brief Free memory from the PACOVR scratch heap.
+ *
+ * @param p Pointer to the PACOVR instance.
+ * @param ptr Pointer to the memory to free.
+ */
+void pacovr_scratch_free(struct pacovr *p, void *ptr)
+{
+	if (!p || !ptr)
+		return;
+
+	sys_heap_free(&p->scratch.heap, ptr);
+}
+
+/**
+ * @brief Allocate memory from the PACOVR batch allocator.
+ *
+ * @param p Pointer to the PACOVR instance.
+ * @param size Size of the allocation.
+ * @return void* Pointer to the allocated memory, or NULL on failure.
+ */
+void *pacovr_batch_alloc(struct pacovr *p, size_t size)
+{
+	void *ptr;
+
+	if (!p || !size)
+		return NULL;
+
+	/* align up size to 4 bytes - force aligned loads and stores */
+	size = ALIGN_UP(size, 4);
+
+	/* check we have enough batch space left */
+	if (p->batch_used + size > p->batch_size) {
+		LOG_ERR("error: pacovr batch alloc failed for %d bytes, only %d bytes free",
+			size, p->batch_size - p->batch_used);
+		return NULL;
+	}
+
+	/* allocate memory */
+	ptr = p->batch_ptr;
+	p->batch_ptr += size;
+	p->batch_used += size;
```

Comment:
are "batch" and "scratch" allocator names commonly used? I found a couple of references to scratch allocators, e.g. https://stackoverflow.com/questions/74956010/what-are-the-differences-between-block-stack-and-scratch-allocators#75044083 but there it's defined as "once allocated, never freed," similar to this your batch allocator?